From 3b6b3791edb551ea9963486c78e5b3c4bd721acd Mon Sep 17 00:00:00 2001
From: Huiwen He <hehuiwen@kylinos.cn>
Date: Wed, 18 Feb 2026 00:30:55 +0800
Subject: [PATCH 5/7] smb/client: refactor ntstatus_to_dos() and
 cifs_print_status()

Modify ntstatus_to_dos() to return a 'const struct nt_err_code_struct *'
instead of using output parameters. This allows callers to access all
fields of the mapping entry directly.

Update cifs_print_status() to accept the mapping struct pointer instead
of a status code, avoiding redundant lookups.

Update map_smb_to_linux_error() to initialize the mapping pointer early
and use the new pointer-based interfaces.

Signed-off-by: Huiwen He <hehuiwen@kylinos.cn>
Reviewed-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/client/smb1maperror.c | 54 ++++++++++++++++--------------------
 1 file changed, 24 insertions(+), 30 deletions(-)

diff --git a/fs/smb/client/smb1maperror.c b/fs/smb/client/smb1maperror.c
index 645c7aca2b55..bf1d2fc6e511 100644
--- a/fs/smb/client/smb1maperror.c
+++ b/fs/smb/client/smb1maperror.c
@@ -38,42 +38,27 @@ static const struct nt_err_code_struct ntstatus_to_dos_map[] = {
 	{0, 0, 0, NULL}
 };
 
-/*****************************************************************************
- Print an error message from the status code
- *****************************************************************************/
-static void
-cifs_print_status(__u32 status_code)
+static const struct nt_err_code_struct *
+ntstatus_to_dos(__u32 ntstatus)
 {
 	int i;
 
 	for (i = 0; ntstatus_to_dos_map[i].nt_errstr; i++) {
-		if (ntstatus_to_dos_map[i].ntstatus == status_code) {
-			pr_notice("Status code returned 0x%08x %s\n",
-				  status_code, ntstatus_to_dos_map[i].nt_errstr);
-			return;
-		}
+		if (ntstatus == ntstatus_to_dos_map[i].ntstatus)
+			return &ntstatus_to_dos_map[i];
 	}
+	return NULL;
 }
 
-
+/*****************************************************************************
+ *Print an error message from the status code
+ *****************************************************************************/
 static void
-ntstatus_to_dos(__u32 ntstatus, __u8 *eclass, __u16 *ecode)
+cifs_print_status(const struct nt_err_code_struct *map)
 {
-	int i;
-	if (ntstatus == 0) {
-		*eclass = 0;
-		*ecode = 0;
-		return;
-	}
-	for (i = 0; ntstatus_to_dos_map[i].ntstatus; i++) {
-		if (ntstatus == ntstatus_to_dos_map[i].ntstatus) {
-			*eclass = ntstatus_to_dos_map[i].dos_class;
-			*ecode = ntstatus_to_dos_map[i].dos_code;
-			return;
-		}
-	}
-	*eclass = ERRHRD;
-	*ecode = ERRgeneral;
+	if (map)
+		pr_notice("Status code returned 0x%08x %s\n",
+			  map->ntstatus, map->nt_errstr);
 }
 
 int
@@ -95,11 +80,20 @@ map_smb_to_linux_error(char *buf, bool logErr)
 		/* translate the newer STATUS codes to old style SMB errors
 		 * and then to POSIX errors */
 		__u32 err = le32_to_cpu(smb->Status.CifsError);
+		const struct nt_err_code_struct *map = ntstatus_to_dos(err);
+
 		if (logErr && (err != (NT_STATUS_MORE_PROCESSING_REQUIRED)))
-			cifs_print_status(err);
+			cifs_print_status(map);
 		else if (cifsFYI & CIFS_RC)
-			cifs_print_status(err);
-		ntstatus_to_dos(err, &smberrclass, &smberrcode);
+			cifs_print_status(map);
+
+		if (map) {
+			smberrclass = map->dos_class;
+			smberrcode = map->dos_code;
+		} else {
+			smberrclass = ERRHRD;
+			smberrcode = ERRgeneral;
+		}
 	} else {
 		smberrclass = smb->Status.DosError.ErrorClass;
 		smberrcode = le16_to_cpu(smb->Status.DosError.Error);
-- 
2.53.0

